### PR TITLE
all: background state committer

### PR DESCRIPTION
**Pull request by rjl493456442**
We can divide the whole block processing into several stages:
- Run transactions
  - Load state
  - Execute in the EVM
  - Store state
- Update changes into trie and rehash
- Commit state and run GC algorithm

The idea behind this PR is: for the commit stage, it can be put in the background and run concurrently with other stages like the pipeline.

### Reuse uncomitted trie
Everytime after rehashing the trie, the "uncommitted" trie actually can be reused for processing the next block.
The rationale of this is all freshly created nodes are referenced in the trie itself. The following blocks can visit the
latest state anyway. 

### Commit scope
A new notion "commit scope" is introduced here. Seems an uncomitted trie can be reused for state retrieval and state rehash, so in order to prevent the dirty nodes are committed multiple times, each phase will be assigned a commit sequence. Whenever an uncommitted trie is reused, the commit sequence is bumped.

This commit sequence can be used to differentiate commit scope.

### Background GC
Instead of committing the dirty nodes into the memory DB, memory GC algo will be applied for each block. It can also be put in the background

### Performance affect
In theory we can cut the "state commit" and "memory GC" time

### TODO
- [x] Run fullsync, ensure the code is correct
- [x] Get some performance numbers

### CODE REVIEW

**Comment by rjl493456442** [comment]
After running for a day, we can see the PR is slightly faster than master(**Exp = Bench03, Master = Bench04**)

截屏2020-07-29上午10 18 44

For more details, the "account commit" and "storage commit" are nuked, "commit" is much faster than the master.
However, the other parts(e.g. execution, trie update ... ) is slightly slower than the master.
截屏2020-07-29上午10 19 53

File: crypto/bls12381/bls12_381.go [hunk 1]
```diff
@@ -119,105 +119,105 @@ var g2One = PointG2{
 */
 
 var frobeniusCoeffs61 = [6]fe2{
-	fe2{
+	{
```
Comment (holiman) [inline crypto/bls12381/bls12_381.go:122]: automated linter change?

Comment (rjl493456442) [reply crypto/bls12381/bls12_381.go:122]: Yes, gofmt

**Comment by rjl493456442** [comment]
The PR is ~26h faster than master

image

**Comment by holiman** [comment]
@rjl493456442 now that we have merged both the snapshotter and the new prefetcher, it might be worth revisiting this -- the numbers look really good. Would you care to rebase this?

**Comment by rjl493456442** [comment]
@holiman of course. I will spend time on it very soon. It's really nice we still have the interest on this.

**Comment by rjl493456442** [comment]
Close it. The state organization will be changed very soon. It's unneeded to do background commit.

**Pull request closed without merging**